### PR TITLE
Explore preprocessing strategy

### PR DESCRIPTION
**Pull request by hitorilabs**
- `np.memmap` loses shape information when you load it in again (this doesn't really produce `npy` files)
  - only strategy that allows you to save a single file that's bigger than RAM
  - one huge file can't really be parallelized when transferring.
- `np.save` retains this shape info, but you can't append to an existing `npy` file - so you neet to split data into batches.
  - even though you could load a file bigger than RAM with mmap'd read, you can only save a file that fits in RAM.

`np.float32` for training data and `np.int64` to match `torch.long` (we're going to use cross entropy)

original iris (1.2 KB training set, 120KB targets)
```
(.venv) bocchi@hitorilabs:~/papers$ ls -la datasets
total 24
drwxr-xr-x 2 bocchi bocchi 4096 May 21 18:50 .
drwxr-xr-x 7 bocchi bocchi 4096 May 21 18:50 ..
-rw-r--r-- 1 bocchi bocchi 1200 May 21 18:50 target_memmap.npy
-rw-r--r-- 1 bocchi bocchi  278 May 21 18:50 target_save.npy
-rw-r--r-- 1 bocchi bocchi 2400 May 21 18:50 train_memmap.npy
-rw-r--r-- 1 bocchi bocchi 2528 May 21 18:50 train_save.npy
```

iris x1000000 w/ `np.repeat` (2.4 GB training set, 1.2 GB targets)
```
(.venv) [bocchi@hitorilabs papers]$ ls -la datasets/iris
total 7031280
drwxr-xr-x 2 bocchi bocchi       4096 May 22 03:34 .
drwxr-xr-x 3 bocchi bocchi       4096 May 22 03:34 ..
-rw-r--r-- 1 bocchi bocchi 1200000000 May 22 04:09 target_memmap.npy
-rw-r--r-- 1 bocchi bocchi 1200000128 May 22 04:09 target_save.npy
-rw-r--r-- 1 bocchi bocchi 2400000000 May 22 04:09 train_memmap.npy
-rw-r--r-- 1 bocchi bocchi 2400000128 May 22 04:09 train_save.npy
```

It looks like `np.save` adds exactly 128 bytes to the file (most likely just the shape metadata?)

### CODE REVIEW

File: preprocessing/save_v_mmap.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        "petal_width": np.float32, 
+        "class": "category",
+    },
+    engine="pyarrow"
+    )
+
+xs, ys = (
+    df.loc[:, df.columns != "class"].to_numpy(), 
+    df["class"].cat.codes.to_numpy()
+)
```
Comment (hitorilabs) [inline preprocessing/save_v_mmap.py]: not sure if this category type is actually efficient, but it's a whole lot simpler than "fitting" a label encoder with sklearn or rolling your own logic to apply. 

this lets you pull out either `cat.codes` or `cat.categories` - although if you were doing a buffered read, you would probably need to [pre-define these columns](https://pandas.pydata.org/docs/user_guide/categorical.html) (since pandas needs to map categories to an index)

File: preprocessing/multi_file_example.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+rows = 0
+cols = 0
+for i, file in enumerate(tqdm(list(source_path.glob(pattern)))):
+    df = pd.read_csv(file, header=None)
+    df_rows, df_cols = df.shape
+    rows += df_rows
+    cols = df_cols
+
+train_file = np.memmap(data_path / "train.memmap", dtype='float32', mode='w+', shape=(rows, cols -1))
+target_file = np.memmap(data_path / "target.memmap", dtype='int64', mode='w+', shape=(rows))
```
Comment (hitorilabs) [inline preprocessing/multi_file_example.py:29]: basically have to accept that I'll go through all the files once just to get the shape - then the end user has to be told (1) `dtype` to read in and (2) the shape. 

Just knowing the columns would be enough to satisfy shape (or you could assume `target count == rows`) and fill the rest of the shape with `-1`

File: preprocessing/multi_file_example.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import pandas as pd
+import numpy as np
+import pathlib
+import os
+from tqdm import tqdm
+
+# one-liner for copying data 
+# for((i=1; i <= 100; i++)); do cp source_iris/iris.data "iris_dataset/iris_${i}.data"; done
```
Comment (hitorilabs) [inline preprocessing/multi_file_example.py:8]: for demonstration purposes

**Comment by hitorilabs** [comment]
https://arrow.apache.org/docs/python/pandas.html#zero-copy-series-conversions
